keps/sig-scheduling/3902-decoupled-taint-manager/README.md

```diff
@@ -140,24 +140,27 @@ The creation and startup of the default taint-manager is performed by `kube-cont
 Kubernetes already has a good coverage of node `No-Execute` eviction. We will add tests only for the changed code.
 
 #####  Unit tests
-- `pkg/controller/taintmanager`: 2023-06-03 - 0% (to add)
+- `pkg/controller/tainteviction`: 2024-09-18 - 81.8% 
   - Enable and disable the new `TaintEvictionController`
   - The new `TaintEvictionController` acts on node taints properly
-- `pkg/controller/apis/config/v1alpha1`: 2023-06-03 - 0%
-  - Test new configuration
-- `pkg/controller/nodelifecycle`: 2023-06-03 - 74.8%
+- `pkg/controller/nodelifecycle`: 2024-09-18 - 68%
```

Comment:
It looks like this coverage went down almost 10%

Can you commit to bring it back up?

Reply:
will try

Reply:
@alculquicondor https://github.com/kubernetes/kubernetes/pull/127956 
- **nodelifecycle:** The coverage has now been increased to cover 69.3% of the statements VS the current 64.9% of statements coverage 
- **scheduler** The coverage has now been increased to cover 92.6% of the statements VS the current 85.3% of statements coverage

Some callers are not easy to increase test coverage, i.e.

- `Run` - 0%
- `doNodeProcessingPassWorker` - 0%
- `doPodProcessingWorker`  - 0%

![image](https://github.com/user-attachments/assets/92bdd533-2a6d-4569-aa7f-7df9fcaaf410)

Reply:
unit tests should be able to do a call to `Run` once. See for example the job controller https://github.com/kubernetes/kubernetes/blob/d9c46d8ecb1ede9be30545c9803e17682fcc4b50/pkg/controller/job/job_controller_test.go#L1325

Reply:
Please add a sentence committing to increasing the coverage. We can target 80% coverage.

Reply:
added.

Reply:
> unit tests should be able to do a call to Run once. See for example the job controller


There is the key difference between `Run` and `manager.syncJob`. The former is a blocking call, while the latter is not. `syncJob` is a worker function. `Run` is a blocking call that waits for all workers until the context is canceled. so we can not follow the same approach as you mentioned in the link.

A possible solution is to write like this:
```
func TestXXX(t *testing.T) {
    testCases := []struct{}{}
    c := NewController()
    go c.Run(ctx)
    // do some verification here
}
```

However a lot of test cases have been covered by other unit tests. 

- doNodeProcessingPassWorker calls doNoScheduleTaintingPass which is already tested
- doNoExecuteTaintingPass is already tested
- monitorNodeHealth is already tested